### PR TITLE
[WIP] Added Support for `dry-types`

### PR DESCRIPTION
**Pull request by GustavoCaso**
#25 

@AMHOL as we talk about it I have created this PR so we can discuss design decisions, to add this feature to the lib.

Once we are happy with this we can start working on the new dsl for `dry-types`

### CODE REVIEW

File: lib/dry/configurable.rb [hunk 1]
```diff
@@ -94,11 +94,15 @@ def setting(key, *args, &block)
...
       end
 
+      value, preprocessor = extract_preprocessor(value)
+
       _settings << ::Dry::Configurable::Config::Value.new(
         key,
         value || ::Dry::Configurable::Config::Value::NONE,
-        processor || ::Dry::Configurable::Config::DEFAULT_PROCESSOR
+        processor || ::Dry::Configurable::Config::DEFAULT_PROCESSOR,
+        preprocessor
```
Comment (GustavoCaso) [inline lib/dry/configurable.rb]: I don't know if I really like the idea having a `preprocessor`, but for the example it makes is use. So please feel free to comment as much as need it

File: lib/dry/configurable.rb [hunk 2]
```diff
@@ -122,6 +126,10 @@ def _reader_attributes
 
     private
 
+    def extract_preprocessor(value)
+      value.respond_to?(:call) ? [nil, value] : [value, nil]
```
Comment (GustavoCaso) [inline lib/dry/configurable.rb]: I know it doesn't explain much but I will improve the code probably in the `ArgumentParser` class to handle all this logic

**Comment by AMHOL** [comment]
Personally I think we might be better off making dry-types a mandatory dependency, that way it can handle any processing and setting the default value, i.e.

```ruby
class Something
  extend ::Dry::Configurable

  setting :dsn, Types::Strict::String.default('sqlite::memory')
end
```

**Comment by solnic** [comment]
👍 for making dry-types a dependency, it can encapsulate a lot of functionality and it's already a dependency in other dry-rb gems and rom-rb anyway. It would also be another use case and should help in making dry-types more robust and feature-complete.

**Comment by solnic** [comment]
@GustavoCaso could you rebase this / resolve conflicts?

**Comment by GustavoCaso** [comment]
@solnic @AMHOL adding `dry-types` as a dependency maybe is not that trivial since `dry-configurable` is already a dependency in `dry-types`, so what I did is make it work an interface similar to `dry-types` `#call`.

WDYT?

**Comment by solnic** [comment]
We should remove dry-configurable dep from dry-types, it's an overkill. We only use it to set `:namespace`, which we could simply do via `Dry::Core::ClassAttributes` instead of configurable DSL.

**Comment by GustavoCaso** [comment]
We can close this one. The use of dry-types was added with 1caa28e3b405db806aaf1dc09748e1213ec9bd08

**Pull request closed without merging**